src/diffusers/models/controlnet.py

```diff
@@ -0,0 +1,538 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from dataclasses import dataclass
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+import torch
+from torch import nn
+
+from ..configuration_utils import ConfigMixin, register_to_config
+from ..utils import BaseOutput, logging
+from .cross_attention import AttnProcessor
+from .embeddings import GaussianFourierProjection, TimestepEmbedding, Timesteps
+from .modeling_utils import ModelMixin
+from .unet_2d_blocks import (
+    CrossAttnDownBlock2D,
+    DownBlock2D,
+    UNetMidBlock2DCrossAttn,
+    UNetMidBlock2DSimpleCrossAttn,
+    get_down_block,
+)
+
+
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+
+@dataclass
+class ControlNetOutput(BaseOutput):
+    down_block_res_samples: Tuple[torch.Tensor]
+    mid_block_res_sample: torch.Tensor
+
+
+class ControlNetConditioningDefaultEmbedding(nn.Module):
+    """
+    "Stable Diffusion uses a pre-processing method similar to VQ-GAN [11] to convert the entire dataset of 512 × 512
+    images into smaller 64 × 64 “latent images” for stabilized training. This requires ControlNets to convert
+    image-based conditions to 64 × 64 feature space to match the convolution size. We use a tiny network E(·) of four
+    convolution layers with 4 × 4 kernels and 2 × 2 strides (activated by ReLU, channels are 16, 32, 64, 128,
+    initialized with Gaussian weights, trained jointly with the full model) to encode image-space conditions ... into
+    feature maps ..."
+    """
+
+    def __init__(self, conditioning_channels: int, conditioning_embedding_channels: int):
+        super().__init__()
+
+        self.conditioning_embedder = nn.Sequential(
+            nn.Conv2d(conditioning_channels, 16, kernel_size=3, padding=1),
+            nn.SiLU(),
+            nn.Conv2d(16, 16, kernel_size=3, padding=1),
+            nn.SiLU(),
+            nn.Conv2d(16, 32, kernel_size=3, padding=1, stride=2),
+            nn.SiLU(),
+            nn.Conv2d(32, 32, kernel_size=3, padding=1),
+            nn.SiLU(),
+            nn.Conv2d(32, 96, kernel_size=3, padding=1, stride=2),
+            nn.SiLU(),
+            nn.Conv2d(96, 96, kernel_size=3, padding=1),
+            nn.SiLU(),
+            nn.Conv2d(96, 256, kernel_size=3, padding=1, stride=2),
+            nn.SiLU(),
+            zero_module(nn.Conv2d(256, conditioning_embedding_channels, kernel_size=3, padding=1)),
+        )
+
+    def forward(self, conditioning):
+        embedding = self.conditioning_embedder(conditioning)
+        return embedding
+
+
+class ControlNetModel(ModelMixin, ConfigMixin):
+    _supports_gradient_checkpointing = True
+
+    @register_to_config
+    def __init__(
+        self,
```

Comment:
Let's try to not copy all function inputs here, but only those that are need for ControlNet

Reply:
For the following arguments that are marked can we remove them from the init and instead hardcode the result one below to not blow up the init of the controlnets?

Reply:
will do

Reply:
https://github.com/huggingface/diffusers/pull/2407/commits/2b553d941018c49bf23009e3eae5e245ab835038